Extended marks 2: Electric Boogalo (#11356)
neovim/neovim: runtime/doc/api.txt

    call nvim_win_set_option(win, 'winhl', 'Normal:MyHighlight')
>

==============================================================================
Buffer extended marks	                                  *api-extended-marks*

An extended mark represents a buffer annotation that remains logically
stationary even as the buffer changes. They could be used to represent cursors,
folds, misspelled words, and anything else that needs to track a logical
location in the buffer over time.

Example:

> We will set an extmark at row 1, column three.

Reviewer: `1` and `three` are inconsistent

`let g:mark_ns = nvim_create_namespace('myplugin')`

> `let g:mark_id = nvim_buf_set_extmark(0, :g:mark_ns, 0, 1, 3)`

Reviewer: Is `:g:mark_ns` a typo?

Note: the mark id was randomly generated because we used an inital value of 0

> `echo nvim_buf_lookup_mark(0, g:mark_ns, g:mark_id)`

Author: change

> => [1, 1, 3]

Reviewer: Is this example outdated？

    call nvim_win_set_option(win, 'winhl', 'Normal:MyHighlight')
>

==============================================================================

> Eextended marks						   *api-extended-marks*

Reviewer 2: ```suggestion
Extended marks						   *api-extended-marks*
```

An extended mark (extmark) represents a buffer annotation that follows
movements as the buffer changes. They could be used to represent cursors,
folds, misspelled words, and anything else that needs to track a logical
location in the buffer over time.

Example:

We will set an extmark at the first row and third column. As the API is zero-
indexed, use row and column counts 0 and 2:

`let g:mark_ns = nvim_create_namespace('myplugin')`
`let g:mark_id = nvim_buf_set_extmark(0, g:mark_ns, 0, 0, 2)`

Passing in id=0 creates a new mark and returns the id. we can look-up a mark
by its id:

`echo nvim_buf_get_extmark_by_id(0, g:mark_ns, g:mark_id)`

> => [1, 0, 2]

Reviewer: According to `nvim_buf_get_extmark_by_id`, the return value  should be `[0, 2]`.

Author: Thanks, updated.

Or we can look-up all marks in a buffer for our namespace (or by a range):
`echo nvim_buf_get_extmarks(0, g:mark_ns, 0, -1, -1)`
=> [[1, 0, 2]]

Deleting the text all around an extended mark does not remove it. If you want
to remove an extended mark, use the |nvim_buf_del_extmark()| function.

> The namepsace ensures your plugin doesn't have to deal with extmarks created

Reviewer 2: ```suggestion
The namespace ensures your plugin doesn't have to deal with extmarks created
```

by another plugin.

Mark positions changed by an edit will be restored on undo/redo. Creating and

> deleting marks doesn't count as a buffer change on itself, i e new undo

Reviewer 2: ```suggestion
deleting marks doesn't count as a buffer change on itself, i.e. new undo
```

                Return: ~
                    (row, col) tuple

                                                *nvim_buf_get_extmark_by_id()*
nvim_buf_get_extmark_by_id({buffer}, {namespace}, {id})
                Returns position info for a given extmark id

                Parameters: ~
                    {buffer}     The buffer handle
                    {namespace}  a identifier returned previously with
                                 nvim_create_namespace
                    {id}         the extmark id

                Return: ~
                    (row, col) tuple or empty list () if extmark id was absent

                                                     *nvim_buf_get_extmarks()*
nvim_buf_get_extmarks({buffer}, {ns_id}, {start}, {end}, {amount})
                List extmarks in a range (inclusive)

                range ends can be specified as (row, col) tuples, as well as
                extmark ids in the same namespace. In addition, 0 and -1 works
                as shorthands for (0,0) and (-1,-1) respectively, so that all

>                 marks in the buffer can be quieried as:

Reviewer 2: ```suggestion
                marks in the buffer can be queried as:
```

                all_marks = nvim_buf_get_extmarks(0, my_ns, 0, -1, -1)

                If end is a lower position than start, then the range will be
                traversed backwards. This is mostly used with limited amount,
                to be able to get the first marks prior to a given position.

                Parameters: ~
                    {buffer}     The buffer handle

>                     {namespace}  An id returned previously from

Author: wrong parameter names (also other functions)

                                 nvim_create_namespace
                    {lower}      One of: extmark id, (row, col) or 0, -1 for
                                 buffer ends
                    {upper}      One of: extmark id, (row, col) or 0, -1 for
                                 buffer ends
                    {amount}     Maximum number of marks to return or -1 for
                                 all marks found ///

                Return: ~
                    [[nsmark_id, row, col], ...]

                                                      *nvim_buf_set_extmark()*
nvim_buf_set_extmark({buffer}, {ns_id}, {id}, {line}, {col})
                Create or update a namespaced mark at a position

                If an invalid namespace is given, an error will be raised.

                Parameters: ~
                    {buffer}  The buffer handle
                    {ns_id}   a identifier returned previously with
                              nvim_create_namespace

>                     {id}      The extmark's id or 0 for next free id

Reviewer 2: ```suggestion
                    {id}      The extmark's id or 0 for next available id
```

                    {row}     The row to set the extmark to.
                    {col}     The column to set the extmark to.
                    {opts}    Optional parameters. Currently not used.

                Return: ~

>                     the nsmark_id for a new mark, or 0 for an update

Reviewer 2: ```suggestion
                    the namespaced extmark id for a new mark, or 0 for an update
```
